sklearn/metrics/pairwise.py

```diff
@@ -256,6 +284,123 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+    non-missing coordinates then NaN is returned for that pair.
+
+    Read more in the :ref:`User Guide <metrics>`.
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples_1, n_features)
+
+    Y : {array-like, sparse matrix}, shape (n_samples_2, n_features)
+
+    squared : boolean, optional
+        Return squared Euclidean distances.
+
+    missing_values : "NaN" or integer, optional
+        Representation of missing value
+
+    copy : boolean, optional
+        Make and use a deep copy of X and Y (if Y exists)
+
+    Returns
+    -------
+    distances : {array}, shape (n_samples_1, n_samples_2)
+
+    Examples
+    --------
+    >>> from sklearn.metrics.pairwise import masked_euclidean_distances
+    >>> nan = float("NaN")
+    >>> X = [[0, 1], [1, nan]]
+    >>> # distance between rows of X
+    >>> masked_euclidean_distances(X, X)
+    array([[ 0.        ,  1.41421356],
+           [ 1.41421356,  0.        ]])
+
+    >>> # get distance to origin
+    >>> masked_euclidean_distances(X, [[0, 0]])
+    array([[ 1.        ],
+           [ 1.41421356]])
+
+    References
+    ----------
+    * John K. Dixon, "Pattern Recognition with Partly Missing Data",
+      IEEE Transactions on Systems, Man, and Cybernetics, Volume: 9, Issue:
+      10, pp. 617 - 621, Oct. 1979.
+      http://ieeexplore.ieee.org/abstract/document/4310090/
+
+    See also
+    --------
+    paired_distances : distances betweens pairs of elements of X and Y.
+    """
+
+    # NOTE: force_all_finite=False allows not only NaN but also +/- inf
+    X, Y = check_pairwise_arrays(X, Y, accept_sparse=False,
+                                 force_all_finite=False, copy=copy)
+    if (np.any(np.isinf(X)) or
+            (Y is not X and np.any(np.isinf(Y)))):
+        raise ValueError(
+            "+/- Infinite values are not allowed.")
+
+    # Get missing mask for X and Y.T
+    mask_X = _get_mask(X, missing_values)
+
+    YT = Y.T
+    mask_YT = _get_mask(YT, missing_values)
+
+    # Check if any rows have only missing value
+    if np.any(mask_X.sum(axis=1) == X.shape[1])\
+            or (Y is not X and np.any(mask_YT.sum(axis=0) == Y.shape[1])):
+        raise ValueError("One or more rows only contain missing values.")
+
+    # else:
+    if missing_values not in ["NaN", np.nan] and (
+            np.any(np.isnan(X)) or (Y is not X and np.any(np.isnan(Y)))):
+        raise ValueError(
+            "NaN values present but missing_value = {0}".format(
+                missing_values))
+
+    # Get anti-mask and set Y.T's missing to zero
+    NYT = (~mask_YT).astype(np.int32)
+    YT[mask_YT] = 0
+
+    # Get X anti-mask and set X's missing to zero
+    NX = (~mask_X).astype(np.int32)
+    X[mask_X] = 0
+
+    # Calculate distances
+    # The following formula derived by:
+    # Shreya Bhattarai <shreya.bhattarai@gmail.com>
+
+    distances = (X.shape[1] / (np.dot(NX, NYT))) * \
```

Comment:
can we get rid of the trailing ``\``?

Reply:
Maybe some comments would be nice? ``np.dot(NX, NYT)`` is number of non-missing values for a given ``(x, y)`` pair, right? Then we're doing the matrix binomial formula but weighting it by the nonzeros (and at this point I have to get pen and paper).

Reply:
Btw, is ``X *X`` as fast as ``X ** 2``? Ok so ``np.dot(X * X, NYT)`` is doing a masked row-sum where the missing values in Y are masked out.... Yeah some comments would be nice ;)

Reply:
yeah did some tests which suggest X*X tends to be faster (although only slightly)